draft-ietf-quic-recovery.md

```diff
@@ -1509,27 +1516,34 @@ Invoked from loss detection's OnAckReceived and is supplied with the
 newly acked_packets from sent_packets.
 
 ~~~
-   InCongestionRecovery(sent_time):
-     return sent_time <= congestion_recovery_start_time
-
-   OnPacketsAcked(acked_packets):
-     for packet in acked_packets:
-       // Remove from bytes_in_flight.
-       bytes_in_flight -= packet.size
-       if (InCongestionRecovery(packet.time_sent)):
-         // Do not increase congestion window in recovery period.
-         return
-       if (IsAppOrFlowControlLimited()):
-         // Do not increase congestion_window if application
-         // limited or flow control limited.
-         return
-       if (congestion_window < ssthresh):
-         // Slow start.
-         congestion_window += packet.size
-         return
-       // Congestion avoidance.
-       congestion_window += max_datagram_size * acked_packet.size
-           / congestion_window
+  InCongestionRecovery(sent_time):
+    return sent_time <= congestion_recovery_start_time
+
+  OnPacketsAcked(acked_packets):
+    for acked_packet in acked_packets:
+      OnPacketAcked(acked_packet)
+
+  OnPacketAcked(acked_packet):
+    // Remove from bytes_in_flight.
+    bytes_in_flight -= acked_packet.sent_bytes
+
+    // Do not increase congestion_window if application
+    // limited or flow control limited.
+    if (IsAppOrFlowControlLimited())
+      return
+
+    // Do not increase congestion window in recovery period.
+    if (InCongestionRecovery(acked_packet.time_sent)):
+      return
+
+    if (congestion_window < ssthresh):
+      // Slow start.
+      congestion_window += acked_packet.sent_bytes
+    else:
+      // Congestion avoidance.
+      congestion_window += 
+        max_datagram_size * acked_packet.sent_bytes
```

Comment:
While we are talking about this... this formula may not increase cwnd at all when acked_packet.sent_bytes is pretty small and congestion_window is high (meaning, when mss x sent_bytes < cwnd), because this is integer operation in most of the implementation. It may lead to too conversative cwnd growth during congestion avoidance.

One solution is to use RFC3465 (TCP ABC) style congestion avoidance. https://www.rfc-editor.org/rfc/rfc3465.html#section-2.1

If you guys think it's beneficial I'd open a new PR.

Reply:
Let's fix it in another PR.

Reply:
Yes, using integer math can be a problem here, but I expect each implementation to do what make sense for them to avoid rounding issues.